File: ios/view/Components/ContentComponent.swift
import SwiftUI

struct ContentHeader: View {
    
    private let header: String
    
    init(header: String) {
        self.header = header
    }
    
    var body: some View {
        Text(header)
            .modifier(MaterialTheme.typography.h2)
            .padding(.horizontal, 24)
            .padding(.vertical, 6)
    }
}

struct ContentText: View {
    
    private let text: String
    
    init(text: String) {
        self.text = text
    }
    
    var body: some View {
        Text(text)
            .modifier(MaterialTheme.typography.body1)
            .padding(.horizontal, 24)
            .padding(.vertical, 6)
    }
}

struct ContentImage: View {
    
    private let url: String
    
    init(url: String) {
        self.url = url
    }
    
    var body: some View {
        AsyncImage(urlString: url)
            .frame(height: 192)
            .frame(
                minWidth: 0,
                maxWidth: .infinity
            )
            .clipShape(RoundedRectangle(cornerRadius: 4))
            .padding(.horizontal, 24)
            .padding(.vertical, 6)
    }
}

struct ContentMap: View {
    
    private let id: String
    
    init(id: String) {
        self.id = id
    }
    
    var body: some View {
        AsyncImage(urlString: getUrlFromId(id: id))
            .frame(height: 248)
            .frame(
                minWidth: 0,
                maxWidth: .infinity
            )
            .clipShape(RoundedRectangle(cornerRadius: 4))
            .padding(.horizontal, 24)
            .padding(.vertical, 6)
    }
    
    func getUrlFromId(id: String) -> String {
        switch id {
            case "tropical_zone":
                return "./climate_zone_tropical_zone.png"

            case "tropical_and_subtropical_zone":
                 return "./climate_zone_tropical_and_subtropical_zone.png"

            case "temperate_zone":
                 return "./climate_zone_temperate_zone.png"

            case "subtropical_zone":
                 return "./climate_zone_subtropical_zone.png"

            case "polar_and_subpolar_zone":
                 return "./climate_zone_polar_and_subpolar_zone.png"

            default:
                return "./climate_zone_tropical_zone.png"
        }
    }
}

struct ContentFacts: View {
    
    private let facts: [Fact]
    
    init(facts: [Fact]) {
        self.facts = facts
    }
    
    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 16) {
                ForEach(facts, id: \.self) { element in
                    
                    VStack(alignment: .center, spacing: 0) {
                        
                        Image(self.getIconAssetNameFromType(iconType: element.icon))
                            .resizable()
                            .frame(width: 48, height: 48)
                            .rotationEffect(element.icon == FactIconType.WIDTH ? .degrees(90) : .degrees(0))
                            .foregroundColor(MaterialTheme.colors.PRIMARY_VARIABLE)
                        
                        Text(element.text)
                            .modifier(MaterialTheme.typography.h3)
                            .padding(.top, 8)
                            .frame(
                                minWidth: 0,
                                maxWidth: .infinity
                            )
                        
                    }
                    .frame(width: 96, height: 96)
                    .background(MaterialTheme.colors.SECONDARY)
                    .cornerRadius(4)
                    .shadow(
                        color: Color.black.opacity(0.1),
                        radius: 2,
                        x: 1,
                        y: 1
                    )
                    
                }
            }
            .padding([.leading, .bottom, .trailing], 24)
        }
    }
    
    func getIconAssetNameFromType(iconType: FactIconType) -> String {
        switch iconType {
            case FactIconType.HEIGHT:
                return "baseline_height_black_24pt"
            case FactIconType.WIDTH:
                return "baseline_height_black_24pt"
            case FactIconType.THERMOSTAT:
                return "baseline_device_thermostat_black_24pt"
            case FactIconType.TRAVEL_EXPLORE:
                return "baseline_travel_explore_black_24pt"
        }
    }
}

File: android/view/components/ContentComponent.kt
package com.tobiasheymann.ep.view.components

import androidx.compose.foundation.Image
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Card
import androidx.compose.material.Icon
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.rounded.Grass
import androidx.compose.material.icons.rounded.Height
import androidx.compose.material.icons.rounded.Thermostat
import androidx.compose.material.icons.rounded.TravelExplore
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.rotate
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.dp
import coil.compose.rememberImagePainter
import coil.size.Scale
import com.tobiasheymann.ep.R
import com.tobiasheymann.ep.model.Fact
import com.tobiasheymann.ep.model.FactIconType

@Composable
fun ContentHeader(header: String) {
    Text(
        text = header,
        modifier = Modifier.padding(horizontal = 24.dp, vertical = 6.dp),
        style = MaterialTheme.typography.h2
    )
}

@Composable
fun ContentText(text: String) {
    Text(
        text = text,
        modifier = Modifier.padding(horizontal = 24.dp, vertical = 6.dp),
        style = MaterialTheme.typography.body1
    )
}

@Composable
fun ContentImage(url: String) {
    Image(
        painter = rememberImagePainter(
            data = url,
            builder = {
                crossfade(true)
                scale(Scale.FILL)
                placeholder(drawableResId = R.drawable.placeholder_square_large)
            }
        ),
        contentDescription = null,
        contentScale = ContentScale.Crop,
        modifier = Modifier
            .padding(horizontal = 24.dp, vertical = 6.dp)
            .fillMaxWidth()
            .height(192.dp)
            .clip(shape = RoundedCornerShape(4.dp))
    )
}

@Composable
fun ContentMap(id: String) {

    when (id) {
        "tropical_zone" -> {
            Image(
                painter = rememberImagePainter(
                    data = "./climate_zone_tropical_zone.png",
                ),
                contentDescription = null,
                modifier = Modifier
                    .padding(24.dp)
                    .fillMaxWidth()
                    .height(224.dp)
            )
        }
        "tropical_and_subtropical_zone" -> {
            Image(
                painter = rememberImagePainter(
                    data = "./climate_zone_tropical_and_subtropical_zone.png",
                ),
                contentDescription = null,
                modifier = Modifier
                    .padding(24.dp)
                    .fillMaxWidth()
                    .height(224.dp)
            )
        }
        "temperate_zone" -> {
            Image(
                painter = rememberImagePainter(
                    data = "./climate_zone_temperate_zone.png",
                ),
                contentDescription = null,
                modifier = Modifier
                    .padding(24.dp)
                    .fillMaxWidth()
                    .height(224.dp)
            )
        }
        "subtropical_zone" -> {
            Image(
                painter = rememberImagePainter(
                    data = "./climate_zone_subtropical_zone.png"
                ),
                contentDescription = null,
                modifier = Modifier
                    .padding(24.dp)
                    .fillMaxWidth()
                    .height(224.dp)
            )
        }
        "polar_and_subpolar_zone" -> {
            Image(
                painter = rememberImagePainter(
                    data = "./climate_zone_polar_and_subpolar_zone.png",
                ),
                contentDescription = null,
                modifier = Modifier
                    .padding(24.dp)
                    .fillMaxWidth()
                    .height(224.dp)
            )
        }
    }
}

@Composable
fun ContentFacts(facts: List<Fact>) {

    LazyRow(
        modifier = Modifier.padding(bottom = 24.dp),
        state = rememberLazyListState(),
        horizontalArrangement = Arrangement.spacedBy(16.dp)
    ) {
        itemsIndexed(items = facts) { index, fact ->

            val firstItem = index == 0
            val lastItem = index == facts.size - 1

            Card(
                backgroundColor = MaterialTheme.colors.secondary,
                modifier = Modifier
                    .size(width = if (firstItem || lastItem) 120.dp else 96.dp, height = 96.dp)
                    .padding(start = if (firstItem) 24.dp else 0.dp, end = if (lastItem) 24.dp else 0.dp)
            ) {
                Column(
                    verticalArrangement = Arrangement.Center,
                    horizontalAlignment = Alignment.CenterHorizontally
                ) {
                    Icon(
                        imageVector = getIconFrom(fact.icon),
                        contentDescription = null,
                        modifier = Modifier
                            .size(48.dp)
                            .rotate(if (fact.icon == FactIconType.WIDTH) 90F else 0F),
                        tint = MaterialTheme.colors.primaryVariant
                    )
                    Text(
                        text = fact.text,
                        modifier = Modifier.padding(top = 8.dp),
                        style = MaterialTheme.typography.h3
                    )
                }
            }
        }
    }
}

fun getIconFrom(iconType: FactIconType): ImageVector {
    return when (iconType) {
        FactIconType.HEIGHT -> Icons.Rounded.Height
        FactIconType.WIDTH -> Icons.Rounded.Height
        FactIconType.THERMOSTAT -> Icons.Rounded.Thermostat
        FactIconType.TRAVEL_EXPLORE -> Icons.Rounded.TravelExplore
        else -> Icons.Rounded.Grass
    }
}